packages/orchestrator/internal/metrics/sandboxes.go

```diff
@@ -0,0 +1,208 @@
+package metrics
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/exemplar"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/e2b-dev/infra/packages/orchestrator/internal/sandbox"
+	sbxlogger "github.com/e2b-dev/infra/packages/shared/pkg/logger/sandbox"
+	"github.com/e2b-dev/infra/packages/shared/pkg/smap"
+	"github.com/e2b-dev/infra/packages/shared/pkg/telemetry"
+	"github.com/e2b-dev/infra/packages/shared/pkg/utils"
+)
+
+const (
+	sbxMemThresholdPct       = 80
+	sbxCpuThresholdPct       = 80
+	minEnvdVersionForMetrics = "0.1.5"
+	timeoutGetMetrics        = 50 * time.Millisecond
+	metricsParallelismFactor = 5 // Used to calculate number of concurrently sandbox metrics requests
+)
+
+type (
+	GetSandboxMetricsFunc func(ctx context.Context) (*sandbox.Metrics, error)
+)
+
+type SandboxObserver struct {
+	meterExporter  sdkmetric.Exporter
+	registration   metric.Registration
+	exportInterval time.Duration
+
+	sandboxes *smap.Map[*sandbox.Sandbox]
+
+	meter       metric.Meter
+	cpuTotal    metric.Int64ObservableGauge
+	cpuUsed     metric.Float64ObservableGauge
+	memoryTotal metric.Int64ObservableGauge
+	memoryUsed  metric.Int64ObservableGauge
+}
+
+const shiftFromMiBToBytes = 20 // Shift to convert MiB to bytes
+
+func NewSandboxObserver(ctx context.Context, commitSHA, clientID string, sandboxMetricsExportPeriod time.Duration, sandboxes *smap.Map[*sandbox.Sandbox]) (*SandboxObserver, error) {
```

Comment:
this was moved from telemetry/sandbox.go, only `registration, err := so.startObserving()` part was changed